### ISSUE TITLE
[question/feature request] Uploading packages without going through local cache

### ISSUE THREAD

**Issue by franknarf8** [conan-io/conan#3073]
- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
    - Artifact Repository Manager : Artifactory
    - Build Server : Jenkins
    - Targeted Build Architectures : (Win32, Win64, Mac, Linux, and possibly other in the near future)
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Hello,

Package manager enthusiast, big fan of Conan and first time reporter.

With other package managers (e.g. .Net nuget, JS npm/yarn, python wheel and quite a few others), it is possible to create a package (some kind of tarball) within the project's workspace and upload it in a separate step. This use-case is often useful in the context of build systems where you might want to decouple the time when the project is built&packaged&tested and the time when it is pushed to a remote. (e.g. you might want to build flavors of your library in parallel on different Builder and only push if all the packages were successfully built&packaged&tested; i.e. get the packages/artifacts from the different jobs and push them to your Artifact Repository Manager together)

Correct me if I'm wrong, but currently with Conan :
- it is impossible to build a package in a tarball-like format (which hinders passing the package as a single file to different Builders in a build chain; e.g. with Jenkins you can't use the `stash/unstash` Pipeline functions, or `CopyArtifact` plugin to easily drag the package through different validation steps BEFORE uploading it)
- it is impossible to upload a package to an Artifact Repository Manager without sending it through the local cache first (on local developer systems, that's perfectly fine, but on Builders, I usually prefer never altering anything outside the project's workspace, avoiding polluting the global environment)

Simplified build job use case : 
1. New push to Master
2. Build&Test Win32, Win64, Mac, Linux
    2.1 Building&Package&test_package the different flavors in parallel
    2.2 Building&Package&test_package on different machines (with different local caches)
    2.3 (Optional) Run a more elaborate test (perhaps on other machine) based on the newly built package
3. if and only if all the builds succeeded
    3.1 agregate packages from the 'Build&Test' step
    3.2 push them all together
4. Trigger downstream jobs

I haven't found any discussion on the topic, so I'm sorry if this has been discussed already. I would be very interested in hearing your opinions about these combined features requests.

Thank you and have a nice day!

\- Frank

**Comment by lasote** [comment]
First of all, thanks for your interest in package management and Conan:

> 3.1 agregate packages from the 'Build&Test' step

If it means to build something different to a conan package, yes, you can use the `deploy` method and the imports method (and command) to import the artifacts from the dependency tree to a local directory. Read more here: https://docs.conan.io/en/latest/devtools/running_packages.html?highlight=deploy

> 3.2 push them all together

As it is not a Conan package you could use a rpm, deb, docker...etc repository in Artifactory or a generic one using the Artifactory API to push the new bundle.

Does this help? 

Nice to meet you Frank!

**Comment by memsharded** [comment]
> it is impossible to upload a package to an Artifact Repository Manager without sending it through the local cache first (on local developer systems, that's perfectly fine, but on Builders, I usually prefer never altering anything outside the project's workspace, avoiding polluting the global environment)

That is exactly the purpose of teh conan local cache. It is not global. It doesn't change the environment. On Builders you will use a clean local cache for every job. You can even point the CONAN_USER_HOME env variable to point it to a "deps" directory in your project if you want it that way.

**Comment by franknarf8** [comment]
Dear lastote and memshared,

Thank you for your quick replies!

To answer lastote's question :
no, sorry I think did not explain my situation clearly enough, by `Build&Test` I mean building a Conan receipe with different recipes on different Builders (build servers). The `Build` part boils down to `conan build [...]` and the 'Test' part boils down to `conan test test_package [...]` (and/or using the newly built package to make sure some other project builds, without having the project pushed to the Artifact Repository Manager).
By `3.2 push them all together`, I mean pushing different binaries for the same recipe in one-ish step. (You might not have the newly built "Mac" binaries on your Windows Builder Local Cache and vice-versa)

Oh yeah, I'm using in-source packaging.

To answer memshared's message: 
Thanks for that trick; but is this a recommended approach for Builders? Correct me if I'm wrong, but by always setting `$CONAN_USER_HOME` to the project's directory we will pull the same dependencies from the Artifact Repository Manager, even if it was already used/pulled, right? So no more dependency caching.
I do want to use Conan Local Cache, but only for dependencies, not for the package I am currently building as it might not be fully tested yet (although, I could maybe only set the `$CONAN_USER_HOME` to the project directory when I'm exporting a package).


Allow me to illustrate my current use case : 
![syarp-6imtag9qr34f1rlzq 1](https://user-images.githubusercontent.com/6641563/41647520-ea21b820-7476-11e8-8039-1ab281796d91.png)

- All the Red boxes represet a job being run on ***different*** machine for both efficiency and technical reasons.
- The Conan Packages are Built (`conan build [...]` + ~`conan package/export-pkg [...]`) during the `Pull Dependencies + Build + Unit Test` box. (Same recipies, different binaries)
- The Conan Packages are Tested (`conan test test_package [...]` + Some other project specific external tests) during the `Package Tests + External Tests + Integration Tests` box.
- To make sure we publish something good to our Artifact Repository Manager, we want to wait until all the Builds+Tests of all the platforms are done (as there might be a problem in the code or the recipe, and we do not want to upload a recipe that breaks for one of our supported platforms).

Therefore, we need a way to share packages between Jobs/Builders. As sharing a file between different execution contexts is quite easy, it would be nice to be able to create some tarball/zip with a built package.

I also though of publishing the packages to Artifactory under a different `ci/testing` namespace and then pulling them back in other jobs, but this causes the conan local cache to get dirtied with 'untested packages' that may interfere with each other in cases of rebuilds.

I'm open to other suggestions regarding the recommended Build Server orchestration using Conan; I haven't found good documentation on the subject yet :S.

**Comment by memsharded** [comment]
Hi @franknarf8,

First sorry for having this question unattended for that long. The truth is that the activity in this repo has become insane (which is a great thing :) ), and it has become impossible to respond to all comments.

First, I would like to comment that I have learnt from other users that another possibility is to create temporary Artifactory repos. Don't wait to upload things, you can do it asynchronously from all build jobs. Then, if all succesfully finish, then do a copy of those Artifacts (via the Artifactory API), to the final repo, and remove the temporary one. This is both pretty efficient and can also avoid other errors, like a transfer/network error after a successful build.

Then, to update you that there haven't been significant changes in conan in this respect. It is necessary to ``export-pkg`` binaries first to the conan cache before being able to upload them. Even if progress has been made in ``Editable`` packages (that do not live in the cache, and can be directly consumed in user folders), which probably will be released in next 1.12, that will only involve consumption, but upload won't be possible from user space. This is because there are a few things, like the metadata, and package revisions, that do not exist in user space, they are computed when the package is exported into the conan cache.

In any case, setting a ``$CONAN_USER_HOME`` to the current folder (or ``current_folder/conan_cache``, or whatever), is a totally valid use case, if you don't want conan to put files elsewhere.

Please tell if you have any updated relevant feedback regarding this issue. Thanks!

**Comment by michaelmaguire** [comment]
> First, I would like to comment that I have learnt from other users that another possibility is to create temporary Artifactory repos. Don't wait to upload things, you can do it asynchronously from all build jobs. Then, if all succesfully finish, then do a copy of those Artifacts (via the Artifactory API), to the final repo, and remove the temporary one. This is both pretty efficient and can also avoid other errors, like a transfer/network error after a successful build.

Just catching up on this thread now as it suddenly became relevant to us ;-) Requiring the use of a temporary Artifactory repo here feels a little onerous -- we're also struggling right now to understand whether it's possible to get granular enough Artifactory permissions that will allow our build to do this without requiring that our company's Artifactory maintainers grant us Admin privileges (which they won't do).

It would be really nice if it were possible to easily Jenkins `stash` a built conan package and unstash it on another build machine.

Can we re-open this ticket for a bit more discussion?

**Comment by memsharded** [comment]
Hi @michaelmaguire 

I am re-reading this issue, and if I understood correctly, your use case is different from the one described above, which is basically not having to put packages in the conan cache, while your request seems to be able to "stash" a built package (from the conan cache) and "unstash" it in another different conan cache, in another machine.

The conan cache is pure file-based, so you might be able to stash and unstash part of it (modulo a couple of minor caveats like lock files and making sure the remotes are called the same). In any case, if this is the use case, please close this one and create a new issue with a better title. Thanks!

**Comment by michaelmaguire** [comment]
One thing I was trying to point out that the solution of:

> First, I would like to comment that I have learnt from other users that another possibility is to create temporary Artifactory repos. 

is not something practical for us in our organization.

Artifactory does not support fine-grained enough permissions that would allow my team's build server the permission to create a temporary Conan repo, and our company's Artifactory Enterprise maintainers (rightly) won't grant our team the full Admin permission it would require.

**Comment by lasote** [comment]
Would it be a possibility to have another Artifactory instance for development?

**Comment by michaelmaguire** [comment]
I doubt that's something we would get approval to fund and maintain.

**Comment by memsharded** [comment]
Hi all,

Long time since last update:

- Uploading from a user folder to the server is mostly an impossible workflow, because it is critical to know the ``recipe revision`` and the ``package revision`` in order to upload to the server, and those can only be computed if exported to the cache. This shouldn't be a problem, you can perfectly build locally, define a temporary CONAN_HOME (in Conan 2.0), and ``conan export-pkg`` to that temporary cache and upload it from there, without affecting any other cache or server.
- The second use case from @michaelmaguire is the one in https://github.com/conan-io/conan/issues/10740, that is planned for the 2.0 roadmap

Lets close this ticket and continue conversation in https://github.com/conan-io/conan/issues/10740 for the CI system transfer of packages.
